benchmarks/nemo/nemolite2d/manual_versions/psykal_sycl/time_step_sycl.cpp

```diff
@@ -0,0 +1,686 @@
+        auto area_u_accessor = area_u_buffer.get_access<access::mode::read>(cgh);
+
+        cgh.parallel_for(range<2>(internal_ystop, internal_xstop - 1), [=](id<2> idx){
+            double rtmp1;
+
+            auto jj = idx[0];
+            auto ji = idx[1];
+            if (ji < internal_xstart) return;
+            if (jj < internal_ystart) return;
+
+            if(tmask_accessor[{jj, ji}] + tmask_accessor[{jj, ji + 1}] <= 0){
+                return; // jump over non-computational domain
+            }
+
+            if(tmask_accessor[{jj, ji}] * tmask_accessor[{jj, ji + 1}] > 0){
+                rtmp1 = area_t_accessor[{jj, ji}] * sshn_t_accessor[{jj, ji}] +
+                    area_t_accessor[{jj, ji + 1}] * sshn_t_accessor[{jj, ji + 1}];
+                sshn_u_accessor[{jj, ji}] = 0.5 * rtmp1 / area_u_accessor[{jj, ji}];
+            }else if(tmask_accessor[{jj, ji}] <= 0){
+                sshn_u_accessor[{jj, ji}] = sshn_t_accessor[{jj, ji + 1}];
+            }else if(tmask_accessor[{jj, ji + 1}] <= 0){
+                sshn_u_accessor[{jj, ji}] = sshn_t_accessor[{jj, ji}];
+            }
+        });
+    });
+
+    // Time update kernel (internal domain v points)
+    myqueue.submit([&](handler &cgh){
+        auto sshn_v_accessor = sshn_v_buffer.get_access<access::mode::read_write>(cgh);
+        auto sshn_t_accessor = sshn_t_buffer.get_access<access::mode::read>(cgh);
+        auto tmask_accessor = tmask_buffer.get_access<access::mode::read>(cgh);
+        auto area_t_accessor = area_t_buffer.get_access<access::mode::read>(cgh);
+        auto area_v_accessor = area_v_buffer.get_access<access::mode::read>(cgh);
+
+        cgh.parallel_for(range<2>(internal_ystop - 1, internal_xstop), [=](id<2> idx){
+            double rtmp1;
+            auto jj = idx[0];
+            auto ji = idx[1];
+            if (ji < internal_xstart) return;
+            if (jj < internal_ystart) return;
+
+            if((tmask_accessor[{jj, ji}] + tmask_accessor[{jj + 1, ji}]) <= 0){
+                return; //jump over non-computational domain
+            }
+            if((tmask_accessor[{jj, ji}] * tmask_accessor[{jj + 1, ji}]) > 0){
+                rtmp1 = area_t_accessor[{jj, ji}] * sshn_t_accessor[{jj, ji}] +
+                    area_t_accessor[{jj + 1, ji}] * sshn_t_accessor[{jj + 1, ji}];
+                sshn_v_accessor[{jj, ji}] = 0.5 * rtmp1 / area_v_accessor[{jj, ji}];
+            }else if(tmask_accessor[{jj, ji}] <= 0){
+                sshn_v_accessor[{jj, ji}] = sshn_t_accessor[{jj + 1, ji}];
+            }else if(tmask_accessor[{jj + 1, ji}] <= 0){
+                sshn_v_accessor[{jj, ji}] = sshn_t_accessor[{jj, ji}];
+            }
+        });
+    });
+
+
+#ifdef USE_TIMER
+    TimerStop();
+    TimerReport();
+#endif
+
+}
+
+extern "C" void sycl_read_from_device(void* from, double* to,
+                                        int nx, int ny, int width){
+    // Just add a wait for now (it may not work in accelerators yet)
```

Comment:
Please could you extend this comment a bit to explain whether or not you expect to have to add an explicit copy here in future.

Reply:
I extended the comment.